pilot/pkg/proxy/envoy/v2/eds.go

```diff
@@ -125,14 +238,256 @@ func localityLbEndpointsFromInstances(instances []*model.ServiceInstance) []endp
+	}
+	var discReq *xdsapi.DiscoveryRequest
+	var receiveError error
+	reqChannel := make(chan *xdsapi.DiscoveryRequest, 1)
+	con := &EdsConnection{
+		pushChannel: make(chan bool, 1),
+		PeerAddr:    peerAddr,
+		Clusters:    []string{},
+		Connect:     time.Now(),
+	}
+	var node string
+	go func() {
+		defer close(reqChannel)
+		for {
+			req, err := stream.Recv()
+			if err != nil {
+				log.Errorf("EDS close for client %s %q terminated with errors %v",
+					node, peerAddr, err)
+				for _, c := range con.Clusters {
+					s.removeEdsCon(c, node, con)
+				}
+				if status.Code(err) == codes.Canceled || err == io.EOF {
+					return
+				}
+				receiveError = err
+				return
+			}
+			reqChannel <- req
+		}
+	}()
+	for {
+		// Block until either a request is received or the ticker ticks
+		select {
+		case discReq, ok = <-reqChannel:
+			if !ok {
+				return receiveError
+			}
+			/* Example:
+			 node:<id:"ingress~~istio-ingress-6796c456f4-7zqtm.istio-system~istio-system.svc.cluster.local"
+				cluster:"istio-ingress"
+				build_version:"0/1.6.0-dev//RELEASE" >
+				resource_names:"echosrv.istio-system.svc.cluster.local|http-echo"
+				type_url:"type.googleapis.com/envoy.api.v2.ClusterLoadAssignment"
+			*/
+
+			clusters2 := discReq.GetResourceNames()
+			// Should not change. A node monitors multiple clusters
+			if node == "" && discReq.Node != nil {
+				node = discReq.Node.Id
+			}
+
+			if node == "" {
+				log.Infoa("Missing node ", discReq.String())
+				return errors.New("Missing node " + discReq.String())
+			}
+
+			// This doesn't seem to happen - add a log just in case. Typical requests from envoy
+			// have a single cluster, and
+			if len(con.Clusters) > 0 && len(clusters2) > 0 && con.Clusters[0] != clusters2[0] {
+				log.Infof("EDS: Clusters override %v -> %v %s", con.Clusters, clusters2, discReq.String())
+			}
+			// Doesn't seem to happen - logging in case something changes in envoy
+			if len(clusters2) > 1 {
+				log.Infof("EDS: Batch clusters %v %s", clusters2, discReq.String())
+			}
+			// Given that Pilot holds an eventually consistent data model, Pilot ignores any acknowledgements
+			// from Envoy, whether they indicate ack success or ack failure of Pilot's previous responses.
+			if discReq.ResponseNonce != "" {
+				// TODO: once the deps are updated, log the ErrorCode if set (missing in current version)
+				if edsDebug {
+					log.Infof("EDS: ACK %s %s %s", node, discReq.VersionInfo, con.Clusters)
+				}
+				continue
+			}
+			if len(con.Clusters) > 0 {
+				// Should not happen
+				log.Infof("EDS REQ repeated %s %v/%v %v raw: %s ",
+					node, clusters2, con.Clusters, peerAddr, discReq.String())
+			}
+			// Initial request
+			if edsDebug {
+				log.Infof("EDS REQ %s %v %v raw: %s ",
+					node, clusters2, peerAddr, discReq.String())
+			}
+
+			if len(clusters2) > 0 {
+				con.Clusters = clusters2
+			}
+
+			for _, c := range con.Clusters {
+				s.addEdsCon(c, node, con)
+			}
+
+		case <-con.pushChannel:
```

Comment:
Isn't this case redundant given the code below?

Reply:
Not sure.